pylearn2/datasets/hdf5.py

```diff
@@ -2,338 +2,472 @@
+        else:
+            sources = data_specs[1]
+            data = [self.data[s] for s in sources]
+
+        return FiniteDatasetIterator(self,
+                                     mode(data[0].shape[0],
+                                          batch_size,
+                                          num_batches,
+                                          rng),
+                                     data_specs=data_specs,
+                                     return_tuple=return_tuple,
+                                     convert=convert)
+
+    def _get_canonical_sources(self):
+        """
+        Returns a list. The first element will be the source of `X` if present,
+        or of `topo_view` otherwise. The second element will be the source of
+        `y` if present. This is needed for compatibility with legacy code but
+        you should not rely on it.
+        """
+        source = []
+        if 'X' in self.sources:
+            source.append('features')
+            self.spaces.set_alias('X', 'features')
+            self.data.set_alias('X', 'features')
+        if 'topo_view' in self.sources:
+            source.append('features')
+            self.spaces.set_alias('topo_view', 'features')
+            self.data.set_alias('topo_view', 'features')
+        if 'y' in self.sources:
+            source.append('targets')
+        return source
+
+    def _get_canonical_spaces(self):
+        """
+        Returns a list. The first element will be the space of `X` if present,
+        or of `topo_view` otherwise. The second element will be the space of
+        `y` if present. This is needed for compatibility with legacy code but
+        you should not rely on it.
+        """
+        space = []
+        try:
+            space.append(self.spaces['X'])
+        except KeyError:
+            space.append(self.spaces['topo_view'])
+        try:
+            space.append(self.spaces['y'])
+        except KeyError:
+            pass
+        return space
+
+    def _get_canonical_data(self):
+        """
+        Returns a list. The first element will be the data of `X` if present,
+        or of `topo_view` otherwise. The second element will be the data of
+        `y` if present. This is needed for compatibility with legacy code but
+        you should not rely on it.
+        """
+        data = []
+        try:
+            data.append(self.data['X'])
+        except KeyError:
+            data.append(self.data['topo_view'])
+        try:
+            data.append(self.data['y'])
+        except KeyError:
+            pass
+        return data
 
-class HDF5DatasetIterator(FiniteDatasetIterator):
+    #@wraps(Dataset.get_data_specs, assigned=(), updated=())
+    # not in dataset, but required by iteration.755
+    def get_data_specs(self, source_or_alias=None):
+        """
+            Returns a tuple `(space, source)` related to the provided
+            source_or_alias key, if any, or to the canonical source otherwise.
```

Comment:
If your dataset has data that are not sources, loading all the keys of the dataset would break the compatibility with legacy code. This apply, for instance, to the case in which the default settings for the dataset are stored in the dataset itself.